src/persist/src/indexed/trace.rs

```diff
@@ -122,11 +177,118 @@ impl<K: Data, V: Data> BlobTrace<K, V> {
     ) -> Result<TraceSnapshot<K, V>, Error> {
         let ts_upper = self.ts_upper();
         let mut updates = Vec::with_capacity(self.batches.len());
-        for (_, key) in self.batches.iter() {
-            updates.push(blob.get_trace_batch(key)?);
+        for meta in self.batches.iter() {
+            updates.push(blob.get_trace_batch(&meta.key)?);
         }
         Ok(TraceSnapshot { ts_upper, updates })
     }
+
+    /// Merge two batches into one, forwarding all updates not beyond the current
+    /// `since` frontier to the `since` frontier.
+    fn merge<L: Blob>(
+        &mut self,
+        first: &BlobTraceBatchMeta,
+        second: &BlobTraceBatchMeta,
+        blob: &mut BlobCache<K, V, L>,
+    ) -> Result<BlobTraceBatchMeta, Error> {
+        if first.level != second.level {
+            return Err(format!(
+                "invalid merge of {:?} and {:?}: different compaction levels",
+                first, second
+            )
+            .into());
+        }
+
+        if first.desc.upper() != second.desc.lower() {
+            return Err(format!(
+                "invalid merge of {:?} and {:?}: non contiguous batches",
+                first, second,
+            )
+            .into());
+        }
+
+        let mut updates = vec![];
+
+        updates.extend(blob.get_trace_batch(&first.key)?.updates.iter().cloned());
+        updates.extend(blob.get_trace_batch(&second.key)?.updates.iter().cloned());
+
+        // TODO: use antichain more idiomatically.
+        let since_time = self.since[0];
+        for ((_, _), t, _) in updates.iter_mut() {
+            if *t < since_time {
+                *t = since_time;
+            }
+        }
+
+        // Consolidate
+        differential_dataflow::consolidation::consolidate_updates(&mut updates);
+
+        // Construct a new BlobTraceBatch
+        let desc = Description::new(
+            first.desc.lower().clone(),
+            second.desc.upper().clone(),
+            self.since.clone(),
+        );
+        let new_batch = BlobTraceBatch {
+            desc: desc.clone(),
+            updates,
+        };
+
+        let key = self.new_blob_key();
+        blob.set_trace_batch(key.clone(), new_batch)?;
+
+        // TODO actually clear the unwanted batches from the blob storage
+        Ok(BlobTraceBatchMeta {
+            key,
+            desc,
+            level: first.level + 1,
```

Comment:
nit: crate a `let merged_batch_level = first.level + 1` up by where you assert that the two input levels are the same and use that here. makes it more obvious where this is coming from